File: 209-minimum-size-subarray-sum/209-minimum-size-subarray-sum.rs
use std::cmp::min;

impl Solution {
    pub fn min_sub_array_len(target: i32, nums: Vec<i32>) -> i32 {
        let mut ans = i32::MAX;
        
        let mut nums = nums;
        
        nums.push(0);
        
        let mut n = nums[0];
        nums[0] = 0;
        for i in 1..nums.len(){
            let prev = nums[i];
            nums[i] = nums[i-1] + n;
            n = prev;
        }
        
        
        
        for i in 0..nums.len(){
            let mut left = i + 1;
            let mut right = nums.len();
            
            while (left < right){
                let mid = left + (right - left ) /2;
                
                let sum = nums[mid] - nums[i];
                
                if sum >= target{
                    right = mid;
                    ans = min((mid - i) as i32, ans);
                    
                }else{
                    left = mid+1;
                }
            }
        }
        
        if ans == i32::MAX {
            0
        }else{
            ans
        }
    }
}

File: 209-minimum-size-subarray-sum/209-minimum-size-subarray-sum.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        i, sum, size, m = 0, 0, len(nums), len(nums) + 1
        
        for j in range(0, size):
            sum +=nums[j]
            while sum>=target:
                m = min(m, j - i + 1)
                sum -= nums[i]
                i+=1
                
        return 0 if (m == size + 1) else m
